packages/babel-helpers/src/helpers/decorate.js

```diff
@@ -0,0 +1,604 @@
+    );
+    newElements.push(elementFinishersExtras.element);
+    newElements.push.apply(newElements, elementFinishersExtras.extras);
+    finishers.push.apply(finishers, elementFinishersExtras.finishers);
+  });
+
+  if (!decorators) {
+    return { elements: newElements, finishers: finishers };
+  }
+
+  var result /*: ElementsFinishers */ = _decorateConstructor(
+    newElements,
+    decorators,
+  );
+  finishers.push.apply(finishers, result.finishers);
+  result.finishers = finishers;
+
+  return result;
+}
+
+// AddElementPlacement
+function _addElementPlacement(
+  element /*: ElementDescriptor */,
+  placements /*: Placements */,
+  silent /*: boolean */,
+) {
+  var keys = placements[element.placement];
+  if (!silent && keys.indexOf(element.key) !== -1) {
+    throw new TypeError("Duplicated element (" + element.key + ")");
+  }
+  keys.push(element.key);
+}
+
+// DecorateElement
+function _decorateElement(
+  element /*: ElementDescriptor */,
+  placements /*: Placements */,
+) /*: ElementFinishersExtras */ {
+  var extras /*: ElementDescriptor[] */ = [];
+  var finishers /*: ClassFinisher[] */ = [];
+
+  for (
+    var decorators = element.decorators, i = decorators.length - 1;
+    i >= 0;
+    i--
+  ) {
+    // (inlined) RemoveElementPlacement
+    var keys = placements[element.placement];
+    keys.splice(keys.indexOf(element.key), 1);
+
+    var elementObject /*: ElementObjectInput */ = _fromElementDescriptor(
+      element,
+    );
+    var elementFinisherExtras /*: ElementFinisherExtras */ = _toElementFinisherExtras(
+      (0, decorators[i])(elementObject) /*: ElementObjectOutput */ ||
+        elementObject,
+    );
+
+    element = elementFinisherExtras.element;
+    _addElementPlacement(element, placements);
+
+    if (elementFinisherExtras.finisher) {
+      finishers.push(elementFinisherExtras.finisher);
+    }
+
+    var newExtras /*: ElementDescriptor[] | void */ =
+      elementFinisherExtras.extras;
+    if (newExtras) {
+      for (var j = 0; j < newExtras.length; j++) {
+        _addElementPlacement(newExtras[j], placements);
+      }
+      extras = extras.concat(newExtras);
+    }
+  }
+
+  return { element: element, finishers: finishers, extras: extras };
+}
+
+// DecorateConstructor
+function _decorateConstructor(
+  elements /*: ElementDescriptor[] */,
+  decorators /*: ClassDecorator[] */,
+) /*: ElementsFinishers */ {
+  var finishers /*: ClassFinisher[] */ = [];
+
+  for (var i = decorators.length - 1; i >= 0; i--) {
+    var obj /*: ClassObject */ = _fromClassDescriptor(elements);
+    var elementsAndFinisher /*: ElementsFinisher */ = _toClassDescriptor(
+      (0, decorators[i])(obj) /*: ClassObject */ || obj,
+    );
+
+    if (elementsAndFinisher.finisher !== undefined) {
+      finishers.push(elementsAndFinisher.finisher);
+    }
+
+    if (elementsAndFinisher.elements !== undefined) {
+      elements = elementsAndFinisher.elements;
+
+      for (var j = 0; j < elements.length - 1; j++) {
```

Comment:
How else could it be implemented? Creating a Set instead of the array wouldn't work because we need to check some properties on the obejcts, not the objects themselves.

Reply:
There are only three possible placements, so I think you can implement this as three Sets of keys.